mesonbuild/build.py

```diff
@@ -2417,7 +2429,7 @@ def __init__(self,
                  capture: bool = False,
                  console: bool = False,
                  depend_files: T.Optional[T.Sequence[FileOrString]] = None,
-                 extra_depends: T.Optional[T.Sequence[T.Union[str, SourceOutputs]]] = None,
+                 extra_depends: T.Optional[T.Sequence[T.Union[BuildTarget, CustomTarget, str, File]]] = None,
```

Comment:
This is wrong, You've taken out `CustomTargetIndex`, `GeneratedList`, and `ExtractedObjects`. You want `SourceInputs`, which is `SourceOutupts | str`, which is what you're changing this to.

Reply:
> Among the things you've taken out are `GeneratedList` and `CustomTargetIndex`, which are clearly wrong to take out.

The type of extra_depends was wrong all over the places, made a separate commit to fix it now. Also added CustomTargetIndex which was not accepted by the KwargInfo, fixing that TODO: `# TODO: CustomTargetIndex should be supported here as well`.